### PR TITLE
feat: add sparse rcnn

### PR DESCRIPTION
**Pull request by long8v**

### CODE REVIEW

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+import torch.nn.functional as F
+
+from detectron2.modeling.poolers import ROIPooler, cat
+from detectron2.structures import Boxes
+
+
+_DEFAULT_SCALE_CLAMP = math.log(100000.0 / 16)
+
+
+class DynamicHead(nn.Module):
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:29]: image

image

RoI Align된 feature와 무작위로 초기화되는 proposal features 부분을 합치는 부분 -> DynamicHead로 붙이겠다.

Comment (long8v) [reply SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:29]: 그냥 이 클래스에서 RoIPooler까지 해주넹

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+
+        box_pooler = ROIPooler(
+            output_size=pooler_resolution,
+            scales=pooler_scales,
+            sampling_ratio=sampling_ratio,
+            pooler_type=pooler_type,
+        )
+        return box_pooler
+
+    def forward(self, features, init_bboxes, init_features):
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:92]: forward
- feature
- init_bboxes
- init_features

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+
+        inter_class_logits = []
+        inter_pred_bboxes = []
+
+        bs = len(features[0])
+        bboxes = init_bboxes
+        
+        init_features = init_features[None].repeat(1, bs, 1)
+        proposal_features = init_features.clone()
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:101]: init_features -> proposal features

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+        
+        for rcnn_head in self.head_series:
+            class_logits, pred_bboxes, proposal_features = rcnn_head(features, bboxes, proposal_features, self.box_pooler)
+
+            if self.return_intermediate:
+                inter_class_logits.append(class_logits)
+                inter_pred_bboxes.append(pred_bboxes)
+            bboxes = pred_bboxes.detach()
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:109]: features, proposal_features, box_pooler를 넣어주면 rcnn_head에서 dynamic head를 해주는듯,,? --

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+
+        if self.return_intermediate:
+            return torch.stack(inter_class_logits), torch.stack(inter_pred_bboxes)
+
+        return class_logits[None], pred_bboxes[None]
+
+
+class RCNNHead(nn.Module):
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:117]: RCNNHead!

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+        self.linear1 = nn.Linear(d_model, dim_feedforward)
+        self.dropout = nn.Dropout(dropout)
+        self.linear2 = nn.Linear(dim_feedforward, d_model)
+
+        self.norm1 = nn.LayerNorm(d_model)
+        self.norm2 = nn.LayerNorm(d_model)
+        self.norm3 = nn.LayerNorm(d_model)
+        self.dropout1 = nn.Dropout(dropout)
+        self.dropout2 = nn.Dropout(dropout)
+        self.dropout3 = nn.Dropout(dropout)
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:138]: 여기가 dynamic head인듯!

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+        self.activation = _get_activation_fn(activation)
+
+        # cls.
+        num_cls = cfg.MODEL.SparseRCNN.NUM_CLS
+        cls_module = list()
+        for _ in range(num_cls):
+            cls_module.append(nn.Linear(d_model, d_model, False))
+            cls_module.append(nn.LayerNorm(d_model))
+            cls_module.append(nn.ReLU(inplace=True))
+        self.cls_module = nn.ModuleList(cls_module)
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:149]: num_cls 개수 만큼 linear -> layernorm -> relu

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+
+        # reg.
+        num_reg = cfg.MODEL.SparseRCNN.NUM_REG
+        reg_module = list()
+        for _ in range(num_reg):
+            reg_module.append(nn.Linear(d_model, d_model, False))
+            reg_module.append(nn.LayerNorm(d_model))
+            reg_module.append(nn.ReLU(inplace=True))
+        self.reg_module = nn.ModuleList(reg_module)
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:158]: num_reg 개수만큼

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+        
+        # pred.
+        self.use_focal = cfg.MODEL.SparseRCNN.USE_FOCAL
+        if self.use_focal:
+            self.class_logits = nn.Linear(d_model, num_classes)
+        else:
+            self.class_logits = nn.Linear(d_model, num_classes + 1)
+        self.bboxes_delta = nn.Linear(d_model, 4)
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:166]: 마지막에 classification / bbox regression 하는 부분

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+        self.scale_clamp = scale_clamp
+        self.bbox_weights = bbox_weights
+
+
+    def forward(self, features, bboxes, pro_features, pooler):
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:171]: forward!

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+        """
+        :param bboxes: (N, nr_boxes, 4)
+        :param pro_features: (N, nr_boxes, d_model)
+        """
+
+        N, nr_boxes = bboxes.shape[:2]
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:177]: N개의 bboxes를 받음

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+        
+        # roi_feature.
+        proposal_boxes = list()
+        for b in range(N):
+            proposal_boxes.append(Boxes(bboxes[b]))
+        roi_features = pooler(features, proposal_boxes)            
+        roi_features = roi_features.view(N * nr_boxes, self.d_model, -1).permute(2, 0, 1)
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:184]: bboxes들을 뽑고 pooling 해서 roi_features 뽑음

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+        for b in range(N):
+            proposal_boxes.append(Boxes(bboxes[b]))
+        roi_features = pooler(features, proposal_boxes)            
+        roi_features = roi_features.view(N * nr_boxes, self.d_model, -1).permute(2, 0, 1)        
+
+        # self_att.
+        pro_features = pro_features.view(N, nr_boxes, self.d_model).permute(1, 0, 2)
+        pro_features2 = self.self_attn(pro_features, pro_features, value=pro_features)[0]
+        pro_features = pro_features + self.dropout1(pro_features2)
+        pro_features = self.norm1(pro_features)
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:190]: proposal features들끼리 self-attention. 
proposal features는 지금 N x d

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+
+        # inst_interact.
+        pro_features = pro_features.view(nr_boxes, N, self.d_model).permute(1, 0, 2).reshape(1, N * nr_boxes, self.d_model)
+        pro_features2 = self.inst_interact(pro_features, roi_features)
+        pro_features = pro_features + self.dropout2(pro_features2)
+        obj_features = self.norm2(pro_features)
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:196]: proposal features 들이랑 object features랑 instance interaction하는 부분 (여기가 Dynamic Head)

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+
+        # obj_feature.
+        obj_features2 = self.linear2(self.dropout(self.activation(self.linear1(obj_features))))
+        obj_features = obj_features + self.dropout3(obj_features2)
+        obj_features = self.norm3(obj_features)
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:201]: obj_features는 fcn으로 강화되어 다음 레이어로

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+        reg_feature = fc_feature.clone()
+        for cls_layer in self.cls_module:
+            cls_feature = cls_layer(cls_feature)
+        for reg_layer in self.reg_module:
+            reg_feature = reg_layer(reg_feature)
+        class_logits = self.class_logits(cls_feature)
+        bboxes_deltas = self.bboxes_delta(reg_feature)
+        pred_bboxes = self.apply_deltas(bboxes_deltas, bboxes.view(-1, 4))
+        
+        return class_logits.view(N, nr_boxes, -1), pred_bboxes.view(N, nr_boxes, -1), obj_features
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:214]: class logits / bbox delta 구해서 예측

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+        pred_w = torch.exp(dw) * widths[:, None]
+        pred_h = torch.exp(dh) * heights[:, None]
+
+        pred_boxes = torch.zeros_like(deltas)
+        pred_boxes[:, 0::4] = pred_ctr_x - 0.5 * pred_w  # x1
+        pred_boxes[:, 1::4] = pred_ctr_y - 0.5 * pred_h  # y1
+        pred_boxes[:, 2::4] = pred_ctr_x + 0.5 * pred_w  # x2
+        pred_boxes[:, 3::4] = pred_ctr_y + 0.5 * pred_h  # y2
+
+        return pred_boxes
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:255]: deformable detr iterative refinement처럼 dx, dy, dw, dh 로 하는 부분

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+
+    def __init__(self, cfg, d_model, num_classes, dim_feedforward=2048, nhead=8, dropout=0.1, activation="relu",
+                 scale_clamp: float = _DEFAULT_SCALE_CLAMP, bbox_weights=(2.0, 2.0, 1.0, 1.0)):
+        super().__init__()
+
+        self.d_model = d_model
+
+        # dynamic.
+        self.self_attn = nn.MultiheadAttention(d_model, nhead, dropout=dropout)
+        self.inst_interact = DynamicConv(cfg)
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:127]: 아 여기서 instance interaction할 때 Dynamic Convolution이 쓰임

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+        pred_boxes = torch.zeros_like(deltas)
+        pred_boxes[:, 0::4] = pred_ctr_x - 0.5 * pred_w  # x1
+        pred_boxes[:, 1::4] = pred_ctr_y - 0.5 * pred_h  # y1
+        pred_boxes[:, 2::4] = pred_ctr_x + 0.5 * pred_w  # x2
+        pred_boxes[:, 3::4] = pred_ctr_y + 0.5 * pred_h  # y2
+
+        return pred_boxes
+
+
+class DynamicConv(nn.Module):
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:258]: Dynamic Conv 구현

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 20]
```diff
@@ -0,0 +1,318 @@
...
+
+    def __init__(self, cfg):
+        super().__init__()
+
+        self.hidden_dim = cfg.MODEL.SparseRCNN.HIDDEN_DIM
+        self.dim_dynamic = cfg.MODEL.SparseRCNN.DIM_DYNAMIC
+        self.num_dynamic = cfg.MODEL.SparseRCNN.NUM_DYNAMIC
+        self.num_params = self.hidden_dim * self.dim_dynamic
+        self.dynamic_layer = nn.Linear(self.hidden_dim, self.num_dynamic * self.num_params)
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:267]: dynamic_layer는 hidden_dim -> num_dynamic * num_params로 가는 linear

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 21]
```diff
@@ -0,0 +1,318 @@
...
+        num_output = self.hidden_dim * pooler_resolution ** 2
+        self.out_layer = nn.Linear(num_output, self.hidden_dim)
+        self.norm3 = nn.LayerNorm(self.hidden_dim)
+
+    def forward(self, pro_features, roi_features):
+        '''
+        pro_features: (1,  N * nr_boxes, self.d_model)
+        roi_features: (49, N * nr_boxes, self.d_model)
+        '''
+        features = roi_features.permute(1, 0, 2)
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:284]: roi feature (49, N * nr_boxes, self.d_model) ->  ( N * nr_boxes, 49, self.d_model)

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 22]
```diff
@@ -0,0 +1,318 @@
...
+        parameters = self.dynamic_layer(pro_features).permute(1, 0, 2)
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:285]: pro_features : (49, N * nr_boxes, self.d_model) -> (49, N * nr_boxes, self.num_dynamic * self.num_params)

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 23]
```diff
@@ -0,0 +1,318 @@
...
+
+        param1 = parameters[:, :, :self.num_params].view(-1, self.hidden_dim, self.dim_dynamic)
+        param2 = parameters[:, :, self.num_params:].view(-1, self.dim_dynamic, self.hidden_dim)
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:288]: proposal feature을 num_params 기준으로 두개로 나누고 (hidden_dim, dim_dynmaic), (dim_dynamic, hidden_dim)으로 나눔

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 24]
```diff
@@ -0,0 +1,318 @@
...
+
+        features = torch.bmm(features, param1)
+        features = self.norm1(features)
+        features = self.activation(features)
+
+        features = torch.bmm(features, param2)
+        features = self.norm2(features)
+        features = self.activation(features)
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:296]: param1, param2에 대해 bmm

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py [hunk 25]
```diff
@@ -0,0 +1,318 @@
...
+
+        features = features.flatten(1)
+        features = self.out_layer(features)
+        features = self.norm3(features)
+        features = self.activation(features)
+
+        return features
```
Comment (long8v) [inline SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:303]: linear하나 더 태우고 return.

**Review by long8v — COMMENTED** [review]
23.07.24 모델 아키텍쳐 부분 읽음

**Review by long8v — COMMENTED** [review]
07.31. data 등록하는 부분 + launch 쪽 읽음

File: SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py (hunk 5 above)
Comment (long8v) [reply SparseR-CNN/projects/SparseRCNN/sparsercnn/head.py:117]: RCNNHead: 1 layer 구조 모델
 1) selfattention 
 2) instinstruct (DynamicConv)
 3) FFN
